ocaml/gencert/selfcert.ml

```diff
@@ -0,0 +1,125 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module C = Cmdliner
+module Rsa = Mirage_crypto_pk.Rsa
+module UX = Xapi_stdext_unix.Unixext
+open Rresult (* introduces >>= >>| and R *)
+
+module D = Debug.Make (struct let name = "gencert_selfcert" end)
+
+(** initialize the random number generator at program startup when this
+module is loaded. *)
+let () = Mirage_crypto_rng_unix.initialize ()
+
+(** [write_cert] writes a PKCS12 file to [path]. The typical file
+ extension would be ".pem". It attempts to do that atomically by
+ writing to a temporary file in the same directory first and renaming
+ the file at the end *)
+let write_certs path pkcs12 =
+  let f () =
+    UX.atomic_write_to_file path 0o400 @@ fun fd ->
+    UX.really_write fd pkcs12 0 (String.length pkcs12)
+  in
+  R.trap_exn f () |> R.error_exn_trap_to_msg
+
+let expire_in days =
+  let seconds = days * 24 * 60 * 60 in
+  let start = Ptime_clock.now () in
+  match Ptime.(add_span start @@ Span.of_int_s seconds) with
+  | Some expire ->
+      R.ok (start, expire)
+  | None ->
+      R.error_msgf "can't represent %d as time span" days
+
+let add_dns_names extension = function
+  | [] ->
+      extension
+  | names ->
+      X509.Extension.(
+        add Subject_alt_name
+          (false, X509.General_name.(singleton DNS names))
```

Comment:
The `false` here indicates that the extension is not critical (i.e. can be ignored if the other side doesn't support it).
I think that is fine, they can fall back to the CN.

Reply:
Yes, this is not well documented - I had some comments on the earlier PR and looked at the X509 source code.